core/src/main/scala/com/yahoo/maha/core/query/QueryGenerator.scala

```diff
@@ -260,18 +260,43 @@ object QueryGeneratorHelper {
   }
 }
 
+sealed trait VersionNumber {
+  def number: Int
+}
+case class Version private(number: Int) extends VersionNumber
+object Version {
+  val v0: Version = new Version(0)
+  val v1: Version = new Version(1)
+  val v2: Version = new Version(2)
+  val DEFAULT = v0
+
+  def apply(number: Int) : Version = {
+    new Version(number)
+  }
+}
+
 class QueryGeneratorRegistry {
-  private[this] var queryGeneratorRegistry : Map[Engine, QueryGenerator[_]] = Map.empty
 
-  def isEngineRegistered(engine: Engine): Boolean = synchronized {
-    queryGeneratorRegistry.contains(engine)
+  private[this] var queryGeneratorRegistry : Map[Engine, Map[Version , QueryGenerator[_]]] = Map.empty
+
+  def isEngineRegistered(engine: Engine, version: Option[Version]): Boolean = synchronized {
+    queryGeneratorRegistry.contains(engine) && queryGeneratorRegistry(engine).contains(version.getOrElse(Version.DEFAULT))
+  }
+
+  def register[U <: QueryGenerator[_]](engine: Engine, qg: U, version: Version = Version.DEFAULT) : Unit = synchronized {
+    require(!isEngineRegistered(engine, Option(version)), s"Query generator already defined for engine : $engine and version $version")
+    if (queryGeneratorRegistry.contains(engine)) {
+      queryGeneratorRegistry += (engine -> (Map(version -> qg) ++ queryGeneratorRegistry(engine)))
+    } else {
+      queryGeneratorRegistry += (engine -> Map(version -> qg))
+    }
   }
 
-  def register[U <: QueryGenerator[_]](engine: Engine, qg: U) : Unit = synchronized {
-    require(!queryGeneratorRegistry.contains(engine), s"Query generator already defined for engine : $engine")
-    queryGeneratorRegistry += (engine -> qg)
+  def getGenerator(engine: Engine, version: Option[Version]): Option[QueryGenerator[EngineRequirement]] = {
+    for {
+      versionGeneratorMap <- queryGeneratorRegistry.get(engine)
+      generator <- versionGeneratorMap.get(version.getOrElse(Version.DEFAULT)).asInstanceOf[Option[QueryGenerator[EngineRequirement]]]
```

Comment:
why the casting?

Reply:
Wasn't compiling - "Expression of type Option[QueryGenerator[_]]] doesn't confirm to Option[QueryGenerator[EngineRequirement]]]", fixed it now